process/eventsHandler.go

```diff
@@ -67,42 +68,79 @@ func checkArgs(args ArgsEventsHandler) error {
 	if check.IfNil(args.StatusMetricsHandler) {
 		return common.ErrNilStatusMetricsHandler
 	}
+	if check.IfNil(args.EventsInterceptor) {
+		return ErrNilEventsInterceptor
+	}
 
 	return nil
 }
 
-// HandlePushEvents will handle push events received from observer
-func (eh *eventsHandler) HandlePushEvents(events data.BlockEvents) error {
-	if events.Hash == "" {
-		log.Debug("received empty hash", "event", common.PushLogsAndEvents,
-			"will process", false,
-		)
-		return common.ErrReceivedEmptyEvents
+// HandleSaveBlockEvents will handle save block events received from observer
+func (eh *eventsHandler) HandleSaveBlockEvents(allEvents data.ArgsSaveBlockData) error {
+	blockHash := hex.EncodeToString(allEvents.HeaderHash)
+	shouldProcessPushEvents := eh.shouldProcessSaveBlockEvents(blockHash)
+	if !shouldProcessPushEvents {
+		return nil
 	}
 
-	shouldProcessEvents := true
-	if eh.checkDuplicates {
-		shouldProcessEvents = eh.tryCheckProcessedWithRetry(common.PushLogsAndEvents, events.Hash)
+	eventsData, err := eh.eventsInterceptor.ProcessBlockEvents(&allEvents)
+	if err != nil {
+		return err
 	}
 
-	if !shouldProcessEvents {
-		log.Info("received duplicated events", "event", common.PushLogsAndEvents,
-			"block hash", events.Hash,
+	pushEvents := data.BlockEvents{
+		Hash:      eventsData.Hash,
+		ShardID:   eventsData.Header.GetShardID(),
```

Comment:
is there any chance here to have the header nil, resulting in a panic?

Reply:
no, there is a header check in events interceptor